Add EnableDefaultDeny field to Network Policy

squeed opened the pull request:
(see individual commits for a further breakdown of the work included)
(see the [CFP](https://github.com/cilium/design-cfps/blob/main/cilium/CFP-30572-non-default-deny-policies.md) / [CFP PR](https://github.com/cilium/design-cfps/pull/16) for the design motivation)

This adds a new field, `EnableDefaultDeny`, to CiliumNetworkPolicy and CiliumClusterideNetworkPolicy. It configures whether or not selected endpoints are switched to default-deny mode when selected by the given network policy.

By default, the presence of at least one network policy selecting an endpoint implies that endpoint should drop all non-matching traffic. This makes sense, given that Kubernetes NetworkPolicy does not even have the ability to explicitly deny traffic. If an endpoint has no policies, then all traffic is allowed.

This presents administrators with a difficult situation. In a heterogeneous cluster, they may wish to apply a cluster-wide policy allowing access to a centralized service (e.g. monitoring or DNS). However, if this cluster-wide policy happens to be the first policy that selects a workload, that workload will abruptly find itself with all other traffic being dropped. Hence, this configurable option. If an endpoint has only `enableDefaultDeny: false` policies, then no implicit default-deny rule is created. This means an administrator can create cluster-wide policies without worry that they will disrupt existing workloads.

If an endpoint is subject to at least one `enableDefaultDeny: true` policy, then the existing behavior remains and all traffic not explicitly allowed is dropped.

### Example 1: Allow Prometheus

This policy ensures that Prometheus can access all workloads in the cluster, without creating a default-deny policy that could block existing traffic.

```yaml
kind: CiliumClusterwideNetworkPolicy
spec:
  enableDefaultDeny:
    ingress: false
  endpointSelector: {}
  ingress:
  - fromEndpoints:
    - matchLabels:
        "k8s:io.kubernetes.pod.namespace": monitoring
 ```


### Example 2: Block cloud-provider metadata APIs

This policy ensures that traffic to the "magic" cloud-provider API is blocked, without affecting existing workloads

```yaml
kind: CiliumClusterwideNetworkPolicy
spec:
  enableDefaultDeny:
    egress: false
  endpointSelector: {}
  egressDeny:
  - toCIDR:
    - 169.254.169.254/32
 ```

```release-note
This adds a new policy field, EnableDefaultDeny, which permits the creation of network polices that do not drop non-matching traffic.
```

squeed committed:
policy/api: add EnableDefaultDeny field

This adds a new field, EnableDefaultDeny, to CiliumNetworkPolicy and
CiliumClusterwideNetworkPolicy, that controls whether or not the subject
endpoints of this policy should drop unselected peer traffic.

By default, endpoints are in a default-allow mode. When the first policy
applies to an endpoint, it flips it to a default-deny mode. This option
allows disabling that behavior. If not specified, the existing behavior
remains: for each direction, traffic is allowed unless at least one
policy rule applies to that endpoint. If multiple policies select an
endpoint, then default-deny takes precedence.

It is useful in heterogeneous environments, it may not be desirable to
implicitly drop all non-matching traffic. Consider, for example, the
case where an administrator wishes to ensure a monitoring service can
access all namespaces. If they create a cluster-wide policy allowing
access from the monitoring service, it may create a deny policy where
none was previously; unexpectedly dropping traffic.

See: https://github.com/cilium/design-cfps/blob/main/cilium/CFP-30572-non-default-deny-policies.md

This commit contains only the API changes; a subsequent commit will
introduce the implementation.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
policy: implement non-default-deny policies

This adjust the policy rule generation to take in to account
non-default-deny policies.

As before, an endpoint is normally in a policy-disabled state. If any
policies select this endpoint, then policy is enabled and all
non-allowed traffic is dropped.

If, however, an endpoint is only selected by default-allow policies,
then policy is enabled, but a special wildcard allow policy is inserted.
Since wildcard polcies have very low precedence, this ensures that any
Deny or L7-proxy rules will still take effect.

This commit also fixes tests that incorrectly failed to sanitize rules
before adding to the policy repository, leading to a nil pointer
exception. Production code *always* sanitizes rules before adding.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed commented on pkg/policy/repository.go:
```diff
@@ -789,27 +757,123 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
-		return false, false, nil
+	// If there are no default-Deny ingress rules, but there is a pass + deny ingress rule, we must
+	// add an additional wildcard rule for each direction
+	if !ingress && hasIngressDenyPass {
```
One awkwardness here is that we conflate "policy is enabled" with "there is a default allow rule". In the case where we want to Deny-with-default-allow, we could also "disable policy" but still return deny rules.

For the BPF policy engine, these are identical. However, Envoy consumes these fields as well. So, in the case where we have only deny+pass, I synthesize an explicit allow rule and set policy to enabled rather than "disable policy".

squeed replied:
Note: this comment is no longer accurate; L3L4 rules may have L7 policies attached, so we always synthesize a default allow.

joestringer commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
For posterity I think it could help to have a CFP on https://github.com/cilium/design-cfps, to (a) point out any alternate options that were seriously considered and (b) correlate with another existing proposal (https://github.com/cilium/design-cfps/pull/12) and understand whether we should accept both or one of the options.

joestringer commented on pkg/policy/api/rule.go:
```diff
@@ -93,6 +101,18 @@ type Rule struct {
 	Labels labels.LabelArray `json:"labels,omitempty"`
 
+	// DefaultAction is the action to take if traffic does not match
+	// any selectors. The acepted values are Deny and Pass, and the default
```
nit:
```suggestion
	// any selectors. The accepted values are Deny and Pass, and the default
```

joestringer reviewed (commented):
I didn't review the implementation since the key question seems to be around the user-facing API at the moment. What do you think about putting together a CFP on https://github.com/cilium/design-cfps ? I am trying to encourage a culture of documenting the reasoning and key questions / design decisions so we can record what we're trying to solve, why we solve it this way, and why we didn't solve it other ways. This can also help us to come to general consensus with others in the community. The CFP doesn't have to be super complicated, mostly just the content you already wrote for this PR description. Beyond the content here, I think that a dedicated document can also help us to consider alternate API representations without tying too much to the implementation. 

My initial feeling on this proposal is that this is a solid option for how to express this functionality, but it does also overlap with another CFP that has been out for feedback for a few months so I think it's also worth evaluating how the two CFPs interact (even if we end up deciding that they have different use cases and we want to adopt both).

tamilmani1989 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -51,6 +51,19 @@ spec:
+              defaultAction:
+                default: Deny
+                description: DefaultAction is the action to take if traffic does not
+                  match any selectors. The acepted values are Deny and Pass, and the
```
```suggestion
                  match any selectors. The accepted values are Deny and Pass, and the
```

tamilmani1989 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -2597,6 +2610,19 @@ spec:
+                defaultAction:
+                  default: Deny
+                  description: DefaultAction is the action to take if traffic does
+                    not match any selectors. The acepted values are Deny and Pass,
```
```suggestion
                    not match any selectors. The accepted values are Deny and Pass,
```

tamilmani1989 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumnetworkpolicies.yaml:
```diff
@@ -55,6 +55,19 @@ spec:
+              defaultAction:
+                default: Deny
+                description: DefaultAction is the action to take if traffic does not
+                  match any selectors. The acepted values are Deny and Pass, and the
```
```suggestion
                  match any selectors. The accepted values are Deny and Pass, and the
```

tamilmani1989 commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumnetworkpolicies.yaml:
```diff
@@ -2601,6 +2614,19 @@ spec:
+                defaultAction:
+                  default: Deny
+                  description: DefaultAction is the action to take if traffic does
+                    not match any selectors. The acepted values are Deny and Pass,
```
```suggestion
                    not match any selectors. The accepted values are Deny and Pass,
```

tamilmani1989 commented on pkg/policy/repository.go:
```diff
@@ -789,27 +757,123 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
+	policyMode := GetPolicyEnabled()
+	// If policy enforcement isn't enabled, we do not enable policy
+	// enforcement for the endpoint. We don't care about returning any
+	// rules that match.
+	if policyMode == option.NeverEnforce {
+		return false, false, nil
+	}
```
we have a scenario where it requires some rules to be enforced by default irrespective of policyMode (like accessing cloud infrastructure endpoints should not be allowed at any case)

squeed replied:
I think this is an important feature! However, it's distinct from non-default-deny. We've been talking somewhat about an on-disk policy source; we should try and pull this together in to a CFP.

tamilmani1989 reviewed (commented):
This feature can be availed only with CNP or CCNP. This may not address for vendors who support only k8s network policy. I'm wondering if we can merge passing cmdline changes (#30060) with this PR so that can support both cmdline option and via cnp.

learnitall reviewed (commented):
Thank you for the great PR description and for writing the CFP. I can't comment on the implementation in the second commit, but I have two pieces and read value of the `DefaultAction`, it's not the default action for traffic on the endpoint; it' of feedback here on an initial pass:

1. When appropriate, could you update the documentation to include a page on this field? Especially the use case examples? The content you wrote in the PR and the CFP is great and could be adapted pretty well to be in the docs.
2. The name of the field `DefaultAction` is confusing for me. I associate `DefaultAction` with a configuration that lies outside of a policy, ie, a firewall has a `DefaultAction` for traffic on an interface and rules are applied on top. It feels off when the word "default" is used on a per-policy basis because the default action an agent takes for an endpoint is impacted by all of the `DefaultAction`s of matching policies, meaning, if I go to a policy and read the value of the `DefaultAction`, it's not the default action for traffic on the endpoint, it's how a policy contributes to an agent's decision on what the default action is. I feel there is a way the field could be named to communicate this behavior, something like 'triggerDefaultDeny: true/false`. To me, this is more clear, as it makes the contribution aspect clear. It is also self-documenting, making it easy to understand for folks who aren't familiar with the field: "this policy will or will not trigger a default deny rule for the matched endpoints".

WDYT? Thanks!

squeed committed:
policy/api: change Rule.Sanitize() to pointer receiver

Methods called by Sanitize() may alter the underlying structures. For
example. selectors are aggregated and address families are upper-cased.
However, top-level fields can't be written by Sanitize. In the future,
we'd like to do that. So, give it a pointer receiver.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

squeed committed:
test: add ginkgo default-allow tests

Add some tests that create various mixtures of default-allow and
default-deny policies. It is important that default-deny policies always
take precedence over default-allow. It is also important that Deny rules
take precedence over default-allow.

These need to be integration tests, since they rely on specific
interactions between the userspace and bpf policy engines.

Signed-off-by: Casey Callendrello <cdc@isovalent.com>

learnitall commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1505,6 +1505,28 @@ spec:
+                  multiple policies apply to an endpoint, that endpoint's default
+                  deny will be enabled if any policy requests it. \n This is useful
+                  for creating broad-based network policies that will not cause endpoints
+                  to enter default-deny mode."
```
A couple of wording suggestions on this description, small nits. Let me know what you think:

```suggestion
                description: "EnableDefaultDeny determines whether this policy
                  configures the selected endpoint(s) to have a default deny mode.
                  If enabled, all traffic from the selected endpoint(s) not explicitly allowed by a
                  network policy will be dropped. \n If not specified, the default is
                  true for each traffic direction that has rules, and false otherwise.
                  For example, if a policy only has Ingress or IngressDeny rules,
                  then the default for ingress is true, and egress is false. \n If
                  multiple policies apply to an endpoint, that endpoint's default
                  deny will be enabled if any policy requests it. \n This is useful
                  for creating broad-based network policies that will not cause endpoints
                  to enter default-deny mode."
```

squeed replied:
"from the selected endpoints" would not be correct, since this applies separately for ingress and egress. Otherwise, nice suggestions.

learnitall replied:
Ah good point, that makes sense.

learnitall reviewed (changes requested):
Amazing work! I left a small wording suggestion for the description of the field, but otherwise LGTM.

learnitall reviewed (approved):
Looks good!

christarazi commented on pkg/k8s/apis/cilium.io/client/crds/v2/ciliumclusterwidenetworkpolicies.yaml:
```diff
@@ -1545,6 +1545,28 @@ spec:
+              enableDefaultDeny:
+                description: "EnableDefaultDeny determines whether this policy configures
+                  the subject endpoint(s) to have a default deny mode. If enabled,
+                  this causes all traffic not explicitly allowed by a network policy
+                  to be dropped. \n If not specified, the default is true for each
```
I don't see @learnitall's feedback fully addressed here.

squeed replied:
I mostly incorporated it, is there something you see missing?

christarazi replied:
Looks like my review might have been stale; looks good.

christarazi commented on pkg/policy/repository.go:
```diff
@@ -789,27 +761,117 @@ func (p *Repository) computePolicyEnforcementAndRules(securityIdentity *identity
+		if !ingress || !hasIngressDefaultDeny { // short-circuit len()
+			if len(r.Ingress) > 0 || len(r.IngressDeny) > 0 {
+				ingress = true
+				if *r.EnableDefaultDeny.Ingress {
+					hasIngressDefaultDeny = true
+				}
```
Is it safe to deref, from my understanding, a potentially nil pointer here?

squeed replied:
Good question. We require all rules to have `.Sanitize()` called on them before adding to the policy engine -- this does other required work such as aggregating selectors. So, this cannot be nil.

In fact, I added a check to ensure that `.Sanitize()` was always called, but I had to back it out because tests do not call `.Sanitize()`, *and* they pass invalid rules. #31162.

christarazi commented on test/helpers/wrappers.go:
```diff
@@ -80,7 +80,7 @@ func CurlFail(endpoint string, optionalValues ...interface{}) string {
 	}
 	return fmt.Sprintf(
-		`curl --path-as-is -s -D /dev/stderr --fail --connect-timeout %d --max-time %d %s -w "%s"`,
+		`curl -k --path-as-is -s -D /dev/stderr --fail --connect-timeout %d --max-time %d %s -w "%s"`,
```
What's the need to change this here? I don't see any justification in the commit msg.

squeed replied:
I'm using the apiserver as one of the test endpoints. To `curl https://kubernetes.default/healthz`, you need to accept invalid certificates. Didn't think it was enough of a change to warrant a mention.

christarazi replied:
This hunk is changing behavior of a shared function that's used by many tests, so my comment was to raise awareness on that in case there this change introduces unintended behavior changes in other parts of the test suite.

christarazi commented on test/k8s/manifests/ccnp-allow-apiserver-default-allow.yaml:
```diff
@@ -0,0 +1,17 @@
+      toPorts:
+        - ports:
+          - port: "53"
+
```
Nit: newline

nathanjsweet reviewed (approved):
Overall I like this approach, but as we add more features to the policy engine this approach may not continue to work.

aanm reviewed (approved):
My review was only for the @cilium/github-sec files.
